File: src/Y2024/Day22/main.py
from collections import Counter
from itertools import pairwise, islice, batched
from collections.abc import Iterator
from typing import TextIO
import multiprocessing


def parser(raw_data: TextIO):
    return [*map(int, raw_data.read().splitlines())]


def full_solver(data: list[int]):
    p1 = 0
    p2 = Counter()
    with multiprocessing.Pool() as pool:
        for num, counts in pool.imap_unordered(
            solve, batched(data, 175), chunksize=1,
        ):
            p1 += num
            p2 += counts
    return p1, max(p2.values())


def gen_seq(num: int) -> Iterator[int]:
    yield num
    while True:
        num ^= num << 6 & 0xffffff
        num ^= num >> 5 & 0xffffff
        num ^= num << 11 & 0xffffff
        yield num


def solve(data: list[int]) -> tuple[int, Counter]:
    p1 = 0
    p2 = Counter()
    for seed in data:
        seen = set()
        nums = gen_seq(seed)
        window = [n % 10 for n in islice(nums, None, 4)]
        diffs = [9 + j - i for i, j in pairwise(window)]
        diff = sum(v * 19 ** i for i, v in enumerate(diffs[::-1]))
        prev = window[-1]
        for num in islice(nums, None, 1996):
            price = num % 10
            diff = 9 + price - prev + diff * 19 % 130321
            prev = price
            if diff in seen: continue
            seen.add(diff)
            if price: p2[diff] += price
        p1 += next(nums)
    return p1, p2


if __name__ == '__main__':
    testing = False

    try:
        from Tools.Python.path_stuff import test_path
    except ModuleNotFoundError:
        path = 'input.txt'
    else:
        path = test_path if testing else 'input.txt'

    with open(path, 'r') as file:
        data = parser(file)
    import time
    st = time.perf_counter()
    print(*full_solver(data))
    ed = time.perf_counter()
    print(ed - st)

File: src/Y2024/Day22/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	var data []int
	for scanner.Scan() {
		n, err := strconv.Atoi(scanner.Text())
		if err != nil {
			panic(err)
		}
		data = append(data, n)
	}
    fmt.Println(solve(data))
}

func next(n *int) {
	*n ^= (*n) << 6 & 0xffffff
	*n ^= (*n) >> 5 & 0xffffff
	*n ^= (*n) << 11 & 0xffffff
}

func solve(data []int) (p1, p2 int) {
	p1 = 0
	p2_arr := make([]int, 130321)
	for _, num := range data {
        seen := make([]bool, 130321)
        window := make([]int, 4)
        for i := range window {
            window[3 - i] = num % 10
            next(&num)
        }
        diffs := make([]int, 3)
        for i := range diffs {
            diffs[i] = 9 + window[i] - window[i + 1]
        }
        diff := diffs[0] + diffs[1] * 19 + diffs[2] * 19 * 19
        prev := window[0]
        m := 0
        for i := 0; i < 1996; i++ {
            price := num % 10
            next(&num)
            diff = 9 + price - prev + diff * 19 % 130321
            prev = price
            if seen[diff] {
                continue
            }
            seen[diff] = true
            if price != 0 {
                p2_arr[diff] += price
            }
            m++
        }
        p1 += num
	}
    p2 = 0
    for _, num := range p2_arr {
        p2 = max(p2, num)
    }
    return
}
